File: python/mnki/guard.py
"""
guard() / wrap() for Python: verify a tool call before it runs, against the hosted control plane.
Modes: observe (log only), warn (run, surface evidence), require_approval (escalate only), enforce.

    from mnki import AgentTrustClient, Guard, Denied, ApprovalRequired
    g = Guard(AgentTrustClient("https://mnki.com", api_key="at_verify_…"), agent="invoice-agent")

    @g.wrap("refund.create")
    def refund(customer_id: str, amount: float, currency: str = "EUR"): ...

Local mode (no account): `Guard(world=demo_world(), agent="invoice-agent", action_prefix="")` runs the same
pipeline in-process (`mnki.pipeline`, validated against the conformance vectors).
"""
from __future__ import annotations
import asyncio, functools, hashlib, inspect, json, random, time
from typing import Any, Callable, Optional

from .errors import AgentTrustError
from .identity import AgentIdentity, b64url
from .local import local_verify

Mode = str  # "observe" | "warn" | "require_approval" | "enforce"


class Denied(AgentTrustError):
    """The action was verified and refused. `evidence` is the ✓/⚠/✕ list (fail and warn rows first)."""
    def __init__(self, decision: dict):
        reasons = [r for r in decision.get("reasons", []) if not r.endswith("_verified") and r != "authority_valid"]
        super().__init__(403, "denied", decision); self.decision = decision; self.reasons = decision.get("reasons", [])
        self.args = (f"denied: {', '.join(reasons) or 'policy'}",)
    @property
    def evidence(self) -> list:
        order = {"fail": 0, "warn": 1, "pass": 2, "skipped": 3}
        return sorted(self.decision.get("evidence", []), key=lambda e: order.get(e.get("status"), 9))


class ApprovalRequired(AgentTrustError):
    """A human must decide (or already decided against, or the wait ended)."""
    def __init__(self, approval_id: Optional[str], decision_id: str, status: str, reasons: Optional[list] = None):
        code = {"pending": "approval_required", "rejected": "approval_rejected", "expired": "approval_expired"}.get(status, "approval_timeout")
        super().__init__(202, code, {"approval_id": approval_id, "decision_id": decision_id}); self.approval_id, self.decision_id, self.status, self.reasons = approval_id, decision_id, status, reasons or []


def _default_map(_tool: str, args: dict) -> dict:
    out: dict = {}
    amount = args.get("amount", args.get("value"))
    if isinstance(amount, str) and amount.replace(".", "", 1).isdigit(): amount = float(amount)
    if isinstance(amount, (int, float)) and not isinstance(amount, bool): out["amount"] = amount
    cur = args.get("currency")
    if isinstance(cur, str) and len(cur) == 3 and cur.isupper(): out["currency"] = cur
    for k in ("resource", "customer_id", "customerId", "id"):
        if isinstance(args.get(k), str): out["resource"] = args[k]; break
    if isinstance(args.get("region"), str): out["context"] = {"region": args["region"]}
    return out


def _tags(d: dict) -> list:
    t: list = []; r = d.get("reasons", [])
    if d.get("decision") == "DENY": t.append("would_deny")
    if d.get("decision") == "REQUIRE_APPROVAL": t.append("approval_required")
    if "agent_unknown" in r or any(x.startswith("credential_") for x in r): t.append("missing_identity")
    if "no_delegation" in r or any(x.startswith("delegation_") for x in r): t.append("missing_delegation")
    if any(x in r for x in ("capability_missing", "constraint_violated", "budget_exceeded")): t.append("excess_capability")
    if any(x.startswith("policy:") and not x.endswith(":allow") for x in r): t.append("policy_mismatch")
    return t


class Guard:
    def __init__(self, client: Any = None, agent: str = "", identity: Optional[AgentIdentity] = None, mode: Mode = "enforce", action_prefix: str = "tool:", map_args: Optional[Callable[[str, dict], dict]] = None, on_approval: str = "wait", approval_timeout: float = 300.0, initial_delay: float = 1.0, max_delay: float = 15.0, on_decision: Optional[Callable[[dict], None]] = None, sleep: Callable[[float], None] = time.sleep, world: Optional[dict] = None, now: Optional[Callable[[], Any]] = None):
        if mode not in ("observe", "warn", "require_approval", "enforce"): raise ValueError("mode must be observe | warn | require_approval | enforce")
        if client is None and world is None: raise ValueError("guard needs either `client` (hosted) or `world` (local)")
        if not agent: raise ValueError("agent is required")
        self.client, self.agent, self.identity, self.mode, self.prefix, self.world, self._now = client, agent, identity, mode, action_prefix, world, now
        self.map_args, self.on_approval, self.timeout, self.initial_delay, self.max_delay, self.on_decision, self._sleep = map_args or _default_map, on_approval, approval_timeout, initial_delay, max_delay, on_decision, sleep

    def _input(self, tool: str, args: dict) -> dict:
        mapped = self.map_args(tool, args or {})
        ctx = dict(mapped.pop("context", {}) or {}); ctx.update({"tool": tool, "arguments_hash": b64url(hashlib.sha256(json.dumps(args or {}, sort_keys=True, separators=(",", ":")).encode()).digest())})
        return {"agent": self.agent, "action": f"{self.prefix}{tool}", **mapped, "context": ctx}

    def wait_for_approval(self, approval_id: str) -> str:
        deadline = time.time() + self.timeout; delay = self.initial_delay
        while True:
            s = self.client.approval_status(approval_id); st = s.get("status")
            if st in ("approved", "rejected", "expired"): return st
            if time.time() >= deadline: return "timeout"
            self._sleep(min(max(0.0, deadline - time.time()), delay + random.random() * 0.25)); delay = min(self.max_delay, delay * 2)

    def check(self, tool: str, args: Optional[dict] = None) -> dict:
        """Verify one tool call. Returns the decision when the call may proceed; raises Denied / ApprovalRequired otherwise (mode permitting)."""
        if self.client is not None: d = self.client.verify(self._input(tool, args or {}), identity=self.identity)
        else:
            r = local_verify(self.world, self._input(tool, args or {}), self._now() if self._now else None)
            if not r["ok"]: raise AgentTrustError(400, r["code"])
            d = dict(r["result"]); d.update({"request_id": "local", "decision_id": f"local_{int(time.time() * 1000):x}", "approval_id": "local_pending" if d["decision"] == "REQUIRE_APPROVAL" else None, "latency_ms": 0})
        enforced = self.mode == "enforce" or (self.mode == "require_approval" and d.get("decision") == "REQUIRE_APPROVAL")
        record = {"at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()), "mode": self.mode, "tool": tool, "decision": d.get("decision"), "enforced": enforced, "reasons": d.get("reasons", []), "evidence": d.get("evidence", []), "decision_id": d.get("decision_id"), "approval_id": d.get("approval_id"), "tags": _tags(d)}
        if self.on_decision: self.on_decision(record)
        result = {"allowed": d.get("decision") == "ALLOW", "decision": d.get("decision"), "reasons": d.get("reasons", []), "evidence": d.get("evidence", []), "decision_id": d.get("decision_id"), "approval_id": d.get("approval_id"), "enforced": enforced}
        if d.get("decision") == "ALLOW" or not enforced: result["allowed"] = True; return result
        if d.get("decision") == "DENY": raise Denied(d)
        if self.client is None or self.on_approval == "throw" or not d.get("approval_id"): raise ApprovalRequired(d.get("approval_id"), d.get("decision_id", ""), "pending", d.get("reasons"))
        status = self.wait_for_approval(d["approval_id"])
        if status != "approved": raise ApprovalRequired(d["approval_id"], d.get("decision_id", ""), status, d.get("reasons"))
        try: result["attestation"] = self.client.issue_attestation(d["decision_id"]).get("token")
        except AgentTrustError: pass
        result["allowed"] = True; return result

    def wrap(self, tool: Optional[str] = None):
        """Decorator: the function runs only after `check` resolves. Keyword arguments (or a single dict) are the tool arguments."""
        def deco(fn: Callable):
            name = tool or fn.__name__
            if inspect.iscoroutinefunction(fn):
                @functools.wraps(fn)
                async def aw(*a, **kw):
                    await asyncio.get_running_loop().run_in_executor(None, self.check, name, _args(a, kw)); return await fn(*a, **kw)
                return aw
            @functools.wraps(fn)
            def w(*a, **kw):
                self.check(name, _args(a, kw)); return fn(*a, **kw)
            return w
        return deco


def _args(a: tuple, kw: dict) -> dict:
    if kw: return dict(kw)
    if len(a) == 1 and isinstance(a[0], dict): return a[0]
    return {"args": list(a)} if a else {}


def wrap(client: Any, agent: str, tool: Optional[str] = None, **guard_kw):
    """`@mnki.wrap(client, "invoice-agent", "refund.create")` — a one-line guard for a single tool."""
    return Guard(client, agent, **guard_kw).wrap(tool)

File: packages/sdk-ts/src/guard.ts
/**
 * `guard()` — verify a tool call before it runs. One core, two backends (local world or the hosted control
 * plane), four modes: `observe` (log only), `warn` (run, surface evidence), `require_approval` (escalate where
 * policy says so, never hard-deny), `enforce`. Framework adapters are thin wrappers over `check()` / `wrap()`.
 */
import { bodyHash } from "mnki-verifier";
import { AgentTrustClient, AgentIdentity, AgentTrustError, type Decision, type VerifyInput, type GrantExecution, type GrantRecord, type GrantDecision, type GrantResult } from "./index";
import { ApprovalRequired, Denied, MnkiError, fromApi } from "./errors";
import { localVerify, type LocalWorld } from "./local";

export type GuardMode = "observe" | "warn" | "require_approval" | "enforce";
export interface GuardRecord { at: string; mode: GuardMode; tool: string; action: string; resource?: string; decision: string; enforced: boolean; reasons: string[]; evidence: Decision["evidence"]; decision_id?: string; approval_id?: string | null; tags: string[] }
export interface GuardOptions {
  /** Hosted verification (an `AgentTrustClient` or its options) — omit for local mode. */
  client?: AgentTrustClient | { baseUrl: string; apiKey: string; fetch?: typeof fetch };
  /** Local mode: the world to verify against (see `demoWorld()`). */
  world?: LocalWorld;
  /** Agent id, stable id or name as known to the world / organisation. */
  agent: string;
  /** Signs hosted requests (Agent-Proof). Optional; unsigned requests carry a `request_unsigned` warning. */
  identity?: AgentIdentity;
  mode?: GuardMode;
  /** Prefix for tool actions, e.g. `tool:` → `tool:refund.create`. Use "" when tool names already are actions. */
  actionPrefix?: string;
  /** Map tool arguments to amount / currency / resource / context (defaults read `amount|value`, `currency`, `resource|customer_id|id`). */
  mapArgs?: (tool: string, args: unknown) => Partial<Pick<VerifyInput, "amount" | "currency" | "resource" | "context">>;
  onApproval?: "wait" | "throw";
  approval?: { timeoutMs?: number; initialDelayMs?: number; maxDelayMs?: number };
  /** Called for every decision (observe mode writes its shadow log through this). */
  onDecision?: (r: GuardRecord) => void | Promise<void>;
  now?: () => Date;
}
export interface GuardResult { allowed: boolean; decision: string; reasons: string[]; evidence: Decision["evidence"]; decision_id?: string; approval_id?: string | null; enforced: boolean; attestation?: string }

const defaultMap: NonNullable<GuardOptions["mapArgs"]> = (_tool, args) => {
  const a = (args ?? {}) as Record<string, unknown>; const out: Partial<VerifyInput> = {};
  const amount = typeof a.amount === "number" ? a.amount : typeof a.value === "number" ? a.value : typeof a.amount === "string" && /^\d+(\.\d+)?$/.test(a.amount) ? Number(a.amount) : undefined;
  if (amount !== undefined) out.amount = amount;
  if (typeof a.currency === "string" && /^[A-Z]{3}$/.test(a.currency)) out.currency = a.currency;
  const res = [a.resource, a.customer_id, a.customerId, a.id].find((v) => typeof v === "string") as string | undefined; if (res) out.resource = res;
  const ctx: Record<string, unknown> = {};
  if (typeof a.region === "string") ctx.region = a.region;
  // Receiver-side facts travel as arguments so any mapper sees them: the A2A caller (adapters/a2a) and the §15 effect path.
  if (a.a2a && typeof a.a2a === "object") ctx.a2a = a.a2a;
  if (a.effect_path === "custody" || a.effect_path === "attested" || a.effect_path === "none") ctx.effect_path = a.effect_path;
  if (Object.keys(ctx).length) out.context = ctx;
  return out;
};
const stable = (v: unknown): string => JSON.stringify(v, (_k, x) => (x && typeof x === "object" && !Array.isArray(x) ? Object.fromEntries(Object.keys(x as object).sort().map((k) => [k, (x as Record<string, unknown>)[k]])) : x));
const tagsFor = (d: { decision: string; reasons: string[] }): string[] => {
  const t: string[] = []; const r = d.reasons;
  if (d.decision === "DENY") t.push("would_deny"); if (d.decision === "REQUIRE_APPROVAL") t.push("approval_required");
  if (r.includes("agent_unknown") || r.some((x) => x.startsWith("credential_"))) t.push("missing_identity");
  if (r.includes("no_delegation") || r.some((x) => x.startsWith("delegation_"))) t.push("missing_delegation");
  if (r.includes("capability_missing") || r.includes("constraint_violated") || r.includes("budget_exceeded")) t.push("excess_capability");
  if (r.some((x) => x.startsWith("policy:") && !x.endsWith(":allow"))) t.push("policy_mismatch");
  return t;
};

export class Guard {
  readonly mode: GuardMode; private readonly client: AgentTrustClient | null;
  constructor(private readonly o: GuardOptions) {
    this.mode = o.mode ?? "enforce";
    this.client = o.client ? (o.client instanceof AgentTrustClient ? o.client : new AgentTrustClient(o.client)) : null;
    if (!this.client && !o.world) throw new MnkiError("invalid_request", "guard needs either `client` (hosted) or `world` (local)");
  }
  private input(tool: string, args: unknown): VerifyInput {
    const mapped = (this.o.mapArgs ?? defaultMap)(tool, args);
    return { agent: this.o.agent, action: `${this.o.actionPrefix ?? "tool:"}${tool}`, ...mapped, context: { ...(mapped.context ?? {}), tool, arguments_hash: undefined as unknown as string } };
  }
  /** Verify one tool call. Resolves when the call may proceed; throws `Denied` / `ApprovalRequired` when it may not (mode permitting). */
  async check(tool: string, args: unknown = {}): Promise<GuardResult> {
    const input = this.input(tool, args); input.context = { ...input.context, arguments_hash: await bodyHash(stable(args ?? {})) };
    let d: Decision;
    if (this.client) {
      try { d = await this.client.verify(input, { identity: this.o.identity }); }
      catch (e) { if (e instanceof AgentTrustError) throw fromApi(e.status, e.code, e.detail); throw new MnkiError("network", e instanceof Error ? e.message : String(e)); }
    } else {
      const r = await localVerify(this.o.world!, input, this.o.now?.() ?? new Date());
      if (!r.ok) throw new MnkiError("invalid_request", r.code);
      d = { ...(r.result as unknown as Decision), request_id: "local", decision_id: `local_${Date.now().toString(36)}`, approval_id: r.result.decision === "REQUIRE_APPROVAL" ? "local_pending" : null, latency_ms: 0 };
    }
    const enforced = this.mode === "enforce" || (this.mode === "require_approval" && d.decision === "REQUIRE_APPROVAL");
    const rec: GuardRecord = { at: new Date().toISOString(), mode: this.mode, tool, action: input.action, resource: input.resource, decision: d.decision, enforced, reasons: d.reasons, evidence: d.evidence, decision_id: d.decision_id, approval_id: d.approval_id, tags: tagsFor(d) };
    await this.o.onDecision?.(rec);
    const base: GuardResult = { allowed: d.decision === "ALLOW", decision: d.decision, reasons: d.reasons, evidence: d.evidence, decision_id: d.decision_id, approval_id: d.approval_id, enforced };
    if (d.decision === "ALLOW" || !enforced) return { ...base, allowed: true };
    if (d.decision === "DENY") throw new Denied(d);
    // REQUIRE_APPROVAL
    if (!this.client || (this.o.onApproval ?? "wait") === "throw" || !d.approval_id) throw new ApprovalRequired(d.approval_id, d.decision_id, "pending", d.reasons);
    const status = await this.client.waitForApproval(d.approval_id, this.o.approval);
    if (status !== "approved") throw new ApprovalRequired(d.approval_id, d.decision_id, status, d.reasons);
    const att = await this.client.attestations.issue(d.decision_id).catch(() => null);
    return { ...base, allowed: true, attestation: att?.token };
  }
  /**
   * Access broker: perform one operation on a connection through Agent Trust instead of calling the provider with a
   * credential of your own. Verified, granted once and executed by the broker; waits for a human approval like `check`.
   */
  async access(connection: string, operation: string, params: Record<string, string | number | boolean> = {}): Promise<{ grant: GrantRecord; decision: GrantDecision; result: GrantResult }> {
    if (!this.client) throw new MnkiError("invalid_request", "guard.access needs a hosted client");
    const input = { agent: this.o.agent, connection, operation, params };
    let r: GrantExecution;
    try { r = await this.client.grants.execute(input, { identity: this.o.identity }); }
    catch (e) { if (e instanceof AgentTrustError) throw fromApi(e.status, e.code, e.detail); throw new MnkiError("network", e instanceof Error ? e.message : String(e)); }
    const rec = (d: GrantDecision, enforced: boolean): GuardRecord => ({ at: new Date().toISOString(), mode: this.mode, tool: `${connection}/${operation}`, action: `access:${operation}`, resource: connection, decision: d.decision, enforced, reasons: d.reasons, evidence: d.evidence as Decision["evidence"], decision_id: d.decision_id, approval_id: d.approval_id, tags: tagsFor(d) });
    if (r.status === "executed") { await this.o.onDecision?.(rec(r.decision, true)); return r; }
    await this.o.onDecision?.(rec(r.decision, true));
    if ((this.o.onApproval ?? "wait") === "throw") throw new ApprovalRequired(r.approval_id, r.decision.decision_id, "pending", r.decision.reasons);
    const status = await this.client.waitForApproval(r.approval_id, this.o.approval);
    if (status !== "approved") throw new ApprovalRequired(r.approval_id, r.decision.decision_id, status, r.decision.reasons);
    let again: GrantExecution;
    try { again = await this.client.grants.execute({ ...input, approval_id: r.approval_id }, { identity: this.o.identity }); }
    catch (e) { if (e instanceof AgentTrustError) throw fromApi(e.status, e.code, e.detail); throw new MnkiError("network", e instanceof Error ? e.message : String(e)); }
    if (again.status !== "executed") throw new ApprovalRequired(again.approval_id, again.decision.decision_id, "pending", again.decision.reasons);
    return again;
  }
  /** Wrap a tool implementation: the call runs only after `check` resolves. */
  wrap<A extends unknown[], R>(tool: string, fn: (...args: A) => R | Promise<R>): (...args: A) => Promise<R> {
    return async (...args: A) => { await this.check(tool, args[0]); return fn(...args); };
  }
}
export const createGuard = (o: GuardOptions): Guard => new Guard(o);
